File: import_tiles_to_firestore.py
#!/usr/bin/env python3
import argparse, csv, time, sys
from typing import Optional
import firebase_admin
from firebase_admin import credentials, firestore

BATCH_LIMIT = 450  # jätetään marginaali Firestoren 500/batch rajaan

def init_firestore(project_id: Optional[str], creds_path: Optional[str]):
    if firebase_admin._apps:
        return firestore.client()
    if creds_path:
        cred = credentials.Certificate(creds_path)
        firebase_admin.initialize_app(cred, {"projectId": project_id} if project_id else None)
    else:
        firebase_admin.initialize_app(options={"projectId": project_id} if project_id else None)
    return firestore.client()

def main():
    ap = argparse.ArgumentParser(description="Import Z14 tiles CSV -> Firestore tile_meta_z14/{x_y}")
    ap.add_argument("--tiles-csv", required=True, help="Path to z14_level248_ids_<L2ID>.csv")
    ap.add_argument("--project", help="GCP project id")
    ap.add_argument("--creds", help="Service account JSON")
    ap.add_argument("--collection", default="tile_meta_z14", help="Firestore collection name (default: tile_meta_z14)")
    ap.add_argument("--progress-every", type=int, default=10000, help="Print progress every N rows (default: 10000)")
    ap.add_argument("--dry-run", action="store_true", help="Parse only, don't write")
    args = ap.parse_args()

    db = init_firestore(args.project, args.creds)
    col = db.collection(args.collection)

    total = 0
    written = 0
    skipped_bad = 0
    skipped_not_z14 = 0
    dup_skipped = 0
    seen_ids = set()

    batch = db.batch()
    in_batch = 0
    start = time.time()

    with open(args.tiles_csv, "r", encoding="utf-8") as f:
        reader = csv.DictReader(f)
        required = {"z","x","y","lon","lat","level8_id","level4_id","level2_id"}
        missing = required - set(reader.fieldnames or [])
        if missing:
            print(f"ERROR: CSV missing columns: {sorted(missing)}", file=sys.stderr)
            sys.exit(2)

        for row in reader:
            total += 1
            try:
                z = int(row["z"])
                if z != 14:
                    skipped_not_z14 += 1
                    continue

                x = int(row["x"]); y = int(row["y"])
                lon = float(row["lon"]); lat = float(row["lat"])
                l8 = int(row["level8_id"]) if row["level8_id"] else None
                l4 = int(row["level4_id"]) if row["level4_id"] else None
                l2 = int(row["level2_id"]) if row["level2_id"] else None

                doc_id = f"{x}_{y}"
                if doc_id in seen_ids:
                    dup_skipped += 1
                    continue
                seen_ids.add(doc_id)

                payload = {"l8": l8, "l4": l4, "l2": l2, "lon": lon, "lat": lat}

                if not args.dry_run:
                    ref = col.document(doc_id)
                    batch.set(ref, payload, merge=True)
                    in_batch += 1
                    written += 1

                    if in_batch >= BATCH_LIMIT:
                        batch.commit()
                        elapsed = time.time() - start
                        print(f"Committed {in_batch} (total written {written:,}) "
                              f"| elapsed {elapsed:,.1f}s | rate {written/elapsed:,.1f}/s")
                        batch = db.batch()
                        in_batch = 0

            except Exception as e:
                skipped_bad += 1
                if args.progress_every <= 1000:
                    # Näytä virhe useammin, jos pienempi progress-tahti
                    print(f"SKIP row {total}: {e}", file=sys.stderr)

            if args.progress_every and total % args.progress_every == 0:
                elapsed = time.time() - start
                print(f"[progress] rows={total:,} written={written:,} "
                      f"skipped(z!=14)={skipped_not_z14:,} bad={skipped_bad:,} dups={dup_skipped:,} "
                      f"| {elapsed:,.1f}s @ {(written/elapsed if elapsed>0 else 0):.1f}/s")

    if not args.dry_run and in_batch > 0:
        batch.commit()
        print(f"Committed final {in_batch} (total written {written:,})")

    elapsed = time.time() - start
    print(f"DONE: rows={total:,} written={written:,} skipped(z!=14)={skipped_not_z14:,} "
          f"bad={skipped_bad:,} dups={dup_skipped:,}")
    print(f"Time: {elapsed:,.1f}s, rate: {(written/elapsed if elapsed>0 else 0):.1f}/s")
    print(f"Collection: {args.collection}")

if __name__ == "__main__":
    main()

File: import_tiles_to_firestore.js
#!/usr/bin/env node
/* eslint-disable no-console */
const fs = require("fs");
const path = require("path");
const readline = require("readline");
const admin = require("firebase-admin");

const BATCH_LIMIT = 450; // jätetään marginaali 500/batch-rajasta

// -------- CLI --------
function parseArgs(argv) {
    const out = { collection: "tile_meta_z14", progressEvery: 10000, dryRun: false };
    for (let i = 2; i < argv.length; i++) {
        const k = argv[i];
        const v = argv[i + 1];
        if (k === "--tiles-csv") { out.csv = v; i++; }
        else if (k === "--project") { out.project = v; i++; }
        else if (k === "--creds") { out.creds = v; i++; }
        else if (k === "--collection") { out.collection = v; i++; }
        else if (k === "--progress-every") { out.progressEvery = Number(v); i++; }
        else if (k === "--dry-run") { out.dryRun = true; }
        else { console.error(`Unknown arg: ${k}`); process.exit(2); }
    }
    if (!out.csv) {
        console.error("Usage: node import_tiles_to_firestore.js --tiles-csv <file> [--project <id>] [--creds <sa.json>] [--collection <name>] [--progress-every N] [--dry-run]");
        process.exit(1);
    }
    return out;
}
const args = parseArgs(process.argv);

// -------- Firestore init --------
function initFirestore(projectId, credsPath) {
    if (!admin.apps.length) {
        let cred;
        if (credsPath) {
            const abs = path.resolve(credsPath);
            cred = admin.credential.cert(JSON.parse(fs.readFileSync(abs, "utf8")));
        } else {
            cred = admin.credential.applicationDefault();
        }
        const opt = {};
        if (projectId) opt.projectId = projectId;
        admin.initializeApp({ credential: cred, ...opt });
    }
    return admin.firestore();
}

// -------- CSV lukeminen --------
function splitCsvLine(line) { return line.split(","); }

(async function main() {
    const db = initFirestore(args.project, args.creds);
    const col = db.collection(args.collection);

    if (!fs.existsSync(args.csv)) {
        console.error(`ERROR: file not found: ${args.csv}`);
        process.exit(2);
    }

    const stream = fs.createReadStream(args.csv, { encoding: "utf8" });
    const rl = readline.createInterface({ input: stream, crlfDelay: Infinity });

    let header = null;
    let idx = {};
    let total = 0, written = 0, skippedBad = 0, skippedNotZ14 = 0, dupSkipped = 0;

    const seen = new Set();
    let batch = db.batch();
    let inBatch = 0;
    const t0 = Date.now();

    const required = ["z", "x", "y", "lon", "lat", "place_id", "region_id", "country_id", "p_level", "r_level"];

    for await (const raw of rl) {
        const line = raw.trim();
        if (!line) continue;

        if (!header) {
            header = splitCsvLine(line);
            const missing = required.filter(k => !header.includes(k));
            if (missing.length) {
                console.error(`ERROR: CSV missing columns: ${missing.join(", ")}`);
                process.exit(2);
            }
            header.forEach((h, i) => { idx[h] = i; });
            continue;
        }

        total++;
        try {
            const cols = splitCsvLine(line);

            const z = Number(cols[idx.z]);
            if (z !== 14) { skippedNotZ14++; continue; }

            const x = Number(cols[idx.x]);
            const y = Number(cols[idx.y]);
            const lon = Number(cols[idx.lon]);
            const lat = Number(cols[idx.lat]);
            if ([x, y, lon, lat].some(v => Number.isNaN(v))) throw new Error("NaN in numeric fields");

            const p = cols[idx.place_id] ? Number(cols[idx.place_id]) : null;
            const r = cols[idx.region_id] ? Number(cols[idx.region_id]) : null;
            const l2 = cols[idx.country_id] ? Number(cols[idx.country_id]) : null;
            const p_level = cols[idx.p_level] ? Number(cols[idx.p_level]) : null;
            const r_level = cols[idx.r_level] ? Number(cols[idx.r_level]) : null;

            const docId = `${x}_${y}`;
            if (seen.has(docId)) { dupSkipped++; continue; }
            seen.add(docId);

            const payload = { l2, r, r_level, p, p_level, lon, lat };

            if (!args.dryRun) {
                const ref = col.doc(docId);
                batch.set(ref, payload, { merge: true });
                inBatch++;
                written++;

                if (inBatch >= BATCH_LIMIT) {
                    await batch.commit();
                    const dt = (Date.now() - t0) / 1000;
                    console.log(`Committed ${inBatch} (total written ${written.toLocaleString()}) | elapsed ${dt.toFixed(1)}s | rate ${(written / Math.max(1e-9, dt)).toFixed(1)}/s`);
                    batch = db.batch();
                    inBatch = 0;
                }
            }
        } catch (e) {
            skippedBad++;
            if (args.progressEvery <= 1000) {
                console.error(`SKIP row ${total}: ${(e && e.message) ? e.message : e}`);
            }
        }

        if (args.progressEvery && total % args.progressEvery === 0) {
            const dt = (Date.now() - t0) / 1000;
            console.log(
                `[progress] rows=${total.toLocaleString()} written=${written.toLocaleString()} ` +
                `skipped(z!=14)=${skippedNotZ14.toLocaleString()} bad=${skippedBad.toLocaleString()} dups=${dupSkipped.toLocaleString()} ` +
                `| ${dt.toFixed(1)}s @ ${(written / Math.max(1e-9, dt)).toFixed(1)}/s`
            );
        }
    }

    if (!args.dryRun && inBatch > 0) {
        await batch.commit();
        console.log(`Committed final ${inBatch} (total written ${written.toLocaleString()})`);
    }

    const dt = (Date.now() - t0) / 1000;
    console.log(`DONE: rows=${total.toLocaleString()} written=${written.toLocaleString()} skipped(z!=14)=${skippedNotZ14.toLocaleString()} bad=${skippedBad.toLocaleString()} dups=${dupSkipped.toLocaleString()}`);
    console.log(`Time: ${dt.toFixed(1)}s, rate: ${(written / Math.max(1e-9, dt)).toFixed(1)}/s`);
    console.log(`Collection: ${args.collection}`);
    process.exit(0);
})().catch(err => {
    console.error("ERROR:", err && err.message ? err.message : err);
    process.exit(2);
});
